File: ds1/WpfClient/Controller/ClientController.cs
//
// The MIT License(MIT)
//
// Copyright(c) 2014 Demonsaw LLC
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

namespace DemonSaw.Controller
{
	using DemonSaw.Command.Request;
	using DemonSaw.Component;
	using DemonSaw.Entity;
	using DemonSaw.Utility;
	using System;
	using System.Threading;
	using System.Windows;
	using System.Windows.Controls;
	using System.Windows.Media.Imaging;
	using System.Windows.Threading;

	public sealed class ClientController : EntityController
	{
		#region Properties
		private Timer PingTimer { get; set; }

		private static MainWindow Main { get { return Global.Main; } }
		private static LogController LogController { get { return Global.LogController; } }
		private static MenuController MenuController { get { return Global.MenuController; } }
		private static FileController FileController { get { return Global.FileController; } }
		private static SearchController SearchController { get { return Global.SearchController; } }
		private static DownloadController DownloadController { get { return Global.DownloadController; } }
		private static UploadController UploadController { get { return Global.UploadController; } }
		#endregion

		#region Constructors
		public ClientController(DataGrid view, TextBlock title) : base(view, title) { }
		#endregion

		#region Interface
		public override void Init()
		{
			base.Init();

			PingTimer = new Timer(delegate { Ping(); }, PingTimer, DemonTimeout.Ping, DemonTimeout.Ping);
		}

		public override void Shutdown()
		{
			base.Shutdown();

			if (PingTimer != null)
			{
				PingTimer.Dispose();
				PingTimer = null;
			}
		}

		public override void Update()
		{
			base.Update();

			if (Application.Current != null)
			{
				Application.Current.Dispatcher.BeginInvoke(DispatcherPriority.Background, new Action(delegate
				{
					Title.Text = ((List.Count != 1) ? string.Format("{0:n0} Users", List.Count) : "1 User");
				}));
			}
		}
		#endregion

		#region Utility
		public void Add(ClientComponent client, ServerComponent server, GroupComponent group)
		{
			Entity entity = new Entity();
			entity.Add(client);

			ClientOptionsComponent options = new ClientOptionsComponent();
			entity.Add(options);

			entity.Add(server);

			entity.Add(group);

			FolderListComponent folderList = new FolderListComponent();
			entity.Add(folderList);

			FileMapComponent fileMap = new FileMapComponent();
			entity.Add(fileMap);

			FolderMapComponent folderMap = new FolderMapComponent();
			entity.Add(folderMap);

			ClientMachineComponent machine = new ClientMachineComponent();
			entity.Add<MachineComponent>(machine);

			SessionComponent session = new SessionComponent();
			entity.Add(session);

			FileComponent file = new FileComponent();
			entity.Add(file);

			SearchListComponent search = new SearchListComponent(SearchController.Dispatcher);
			entity.Add(search);

			BrowseComponent browse = new BrowseComponent();
			entity.Add(browse);

			DownloadListComponent download = new DownloadListComponent(DownloadController.Dispatcher);
			entity.Add(download);

			UploadListComponent upload = new UploadListComponent(UploadController.Dispatcher);
			entity.Add(upload);
#if DEBUG
			LogComponent log = new LogComponent(LogController);
			entity.Add(log);
#endif
			Add(entity);
		}

		public void Add()
		{
			Entity entity = new Entity();

			ClientComponent client = new ClientComponent() { Name = Environment.MachineName };
			entity.Add(client);			

			ClientOptionsComponent options = new ClientOptionsComponent();
			entity.Add(options);

			ServerComponent server = new ServerComponent();
			entity.Add(server);

			GroupComponent group = new GroupComponent();
			entity.Add(group);

			FolderListComponent folderList = new FolderListComponent();
			entity.Add(folderList);

			FileMapComponent fileMap = new FileMapComponent();
			entity.Add(fileMap);

			FolderMapComponent folderMap = new FolderMapComponent();
			entity.Add(folderMap);

			ClientMachineComponent machine = new ClientMachineComponent();
			entity.Add<MachineComponent>(machine);

			SessionComponent session = new SessionComponent();
			entity.Add(session);

			FileComponent file = new FileComponent();
			entity.Add(file);

			SearchListComponent search = new SearchListComponent(SearchController.Dispatcher);
			entity.Add(search);

			BrowseComponent browse = new BrowseComponent();
			entity.Add(browse);

			DownloadListComponent download = new DownloadListComponent(DownloadController.Dispatcher);
			entity.Add(download);

			UploadListComponent upload = new UploadListComponent(UploadController.Dispatcher);
			entity.Add(upload);
#if DEBUG
			LogComponent log = new LogComponent(LogController);
			entity.Add(log);
#endif
			Add(entity);
		}

		public override void Add(Entity entity)
		{
			base.Add(entity);

			// Events
			ClientComponent client = entity.Get<ClientComponent>();
			client.ComponentChanged += Update;

			FolderListComponent folders = entity.Get<FolderListComponent>();
			folders.ComponentChanged += FileController.Update;

			ClientMachineComponent machine = (ClientMachineComponent) entity.Get<MachineComponent>();
			machine.ListAdded += SearchController.Add;


			MenuController.Add(entity);
		}

		public override bool Remove(Entity entity)
		{
			MenuController.Remove(entity);

			return base.Remove(entity);
		}

		public void Start(Entity entity)
		{
			ClientMachineComponent machine = entity.Get<ClientMachineComponent>();
			machine.Start();
		}

		public void Stop(Entity entity)
		{
			ClientMachineComponent machine = entity.Get<ClientMachineComponent>();
			machine.Stop();
		}

		public void Restart(Entity entity)
		{
			ClientMachineComponent machine = entity.Get<ClientMachineComponent>();
			machine.Restart();
		}

		public override void Select(Entity entity)
		{
			BitmapImage image = null;

			if (entity != null)
			{
				GroupComponent group = entity.Get<GroupComponent>();
				image = group.Image;

				FileComponent file = entity.Get<FileComponent>();
				FileController.Select(file);
			}
			else
			{
				FileController.Select(entity);
			}

			Main.GroupImage.Source = image;			
		}

		public void Toggle(Entity entity)
		{
			ClientComponent client = entity.Get<ClientComponent>();
			MachineComponent machine = entity.Get<MachineComponent>();
			client.Enabled = !client.Enabled;

			if (client.Enabled)
				machine.Start();
			else
				machine.Stop();
		}
		#endregion

		#region Commands
		public void Info()
		{
			lock (List)
			{
				foreach (Entity entity in List)
				{
					if (entity.Invalid)
						continue;

					try
					{
						// NOTE: No need to close the connection
						InfoRequestCommand command = new InfoRequestCommand(entity);
						command.Execute();
					}
					catch (Exception ex)
					{
						Console.WriteLine(ex.Message);
					}
				}
			}
		}

		private void Ping()
		{
			lock (List)
			{
				foreach (Entity entity in List)
				{
					if (entity.Invalid)
						continue;

					try
					{
						// NOTE: No need to close the connection
						PingRequestCommand command = new PingRequestCommand(entity);
						command.Execute();
					}
					catch (Exception ex)
					{
						Console.WriteLine(ex.Message);
					}
				}
			}			
		}

		#endregion
	}
}

File: ds3/3_ds_mobile/controller/client_controller.cpp
//
// The MIT License(MIT)
//
// Copyright(c) 2014 Demonsaw LLC
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include <cassert>

#include <QTimer>

#include "component/endpoint_component.h"
#include "component/chat_model_component.h"
#include "component/debug_model_component.h"
#include "component/group_model_component.h"
#include "component/client/client_option_component.h"
#include "component/client/client_service_component.h"
#include "component/status/status_component.h"

#include "controller/application_controller.h"
#include "controller/client_controller.h"
#include "model/client_model.h"
#include "thread/thread_info.h"
#include "utility/locale.h"
#include "utility/utf8.h"
#include "utility/value.h"

namespace eja
{
	// Constructor
	client_controller::client_controller(entity_list& list, QObject* parent /*= nullptr*/) : entity_object(parent)
	{
		m_model = new client_model(list, this);

		/*m_proxy = new QSortFilterProxyModel(this);
		m_proxy->setSortRole(Qt::UserRole);
		m_proxy->setSourceModel(m_model);		
		m_proxy->setDynamicSortFilter(true);
		m_proxy->sort(client_model::column::role_name);*/
	}

	// Interface
	void client_controller::add()
	{
		assert(thread_info::main() && has_entity());

		m_model->add(m_entity);

		const auto service = m_entity->get<client_service_component>();
		if (service->invalid())
		{
			m_entity->enable();
			service->start();
		}

		const auto app = application_controller::get();
		app->set_index(m_model->size() - 1);
		
		emit nameChanged();
		emit sizeChanged();
	}

	void client_controller::remove(const int row)
	{
		assert(thread_info::main() && has_entity());

		const auto entity = m_model->get_entity(row);

		if (entity)
		{
			const auto chat_model = entity->get<chat_model_component>();
			chat_model->reset();

			const auto debug_model = entity->get<debug_model_component>();
			debug_model->reset();

			const auto group_model = entity->get<group_model_component>();
			group_model->reset();

			if (entity->enabled())
			{
				const auto service = entity->get<client_service_component>();
				if (service->valid())
					service->async_quit();
			}

			m_model->removeRow(row);

			emit sizeChanged();
		}
	}

	void client_controller::update()
	{
		assert(thread_info::main() && has_entity());

		m_model->update();

		emit sizeChanged();
	}

	void client_controller::restart()
	{
		assert(thread_info::main() && has_entity());

		// Spam
		if (!m_enabled)
			return;

		const auto service = m_entity->get<client_service_component>();
		if (service->valid())
			service->restart();

		// Spam
		m_enabled = false;
		QTimer::singleShot(ui::enter, [this]() { m_enabled = true; });
	}

	void client_controller::clear()
	{
		assert(thread_info::main() && has_entity());

		m_model->clear();

		emit sizeChanged();
	}

	void client_controller::reset()
	{
		assert(has_entity());

		m_model->reset();

		emit sizeChanged();
	}

	// Utility
	bool client_controller::empty() const
	{
		return m_model->empty(); 
	}

	void client_controller::sort()
	{
		m_model->sort();
	}

	// Set
	void client_controller::set_entity(const entity::ptr entity)
	{
		entity_object::set_entity(entity);

		m_model->update();

		emit nameChanged();
		emit colorChanged();
		emit text_colorChanged();
		emit statusChanged();
		emit sizeChanged();
		emit flashChanged();
	}

	void client_controller::set_name(const QString& qname)
	{
		const auto endpoint = m_entity->get<endpoint_component>();
		const auto name = qname.trimmed().toStdString();
		endpoint->set_name(name);
		
		// Model
		application_controller::set_initial(m_entity);

		m_model->update(m_entity);

		emit nameChanged();
	}

	void client_controller::set_color(const QColor& color)
	{
		const auto endpoint = m_entity->get<endpoint_component>();
		endpoint->set_color(color.rgba());

		m_model->update(m_entity);

		emit colorChanged();
		emit text_colorChanged();
	}

	void client_controller::set_flash(const bool flash)
	{
		const auto option = m_entity->get<client_option_component>();
		option->set_flash(flash);

		emit flashChanged();
	}

	// Get
	size_t client_controller::get_size() const 
	{ 
		return m_model->size(); 
	}

	QString client_controller::get_name() const
	{
		const auto endpoint = m_entity->get<endpoint_component>();
		return QString::fromStdString(endpoint->get_name());
	}

	QColor client_controller::get_color() const
	{
		QColor qcolor;
		const auto endpoint = m_entity->get<endpoint_component>();
		qcolor.setRgba(endpoint->get_color());
		return qcolor;
	}

	QColor client_controller::get_text_color() const
	{
		const auto endpoint = m_entity->get<endpoint_component>();
		const double value = ((0.299 * endpoint->get_red()) + (0.587 * endpoint->get_green()) + 0.114 * endpoint->get_blue()) / 255.0;
		return (value <= 0.5) ? QColor(0xffffffff) : QColor(0xff484840);
	}

	QVector4D client_controller::get_color_hsva() const
	{
		QColor qcolor = get_color();
		QVector4D hsva = QVector4D(qcolor.hsvHueF(), qcolor.hsvSaturationF(), qcolor.valueF(), qcolor.alphaF());
		return hsva;
	}

	QColor client_controller::get_status() const
	{
		QColor color;
		const auto status = m_entity->get<status_component>();
		color.setRgba(status->get_color());
		return color;
	}

	bool client_controller::has_flash() const
	{
		const auto option = m_entity->get<client_option_component>();
		return option->has_flash();
	}
}
